File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        ListNode* step1 = head;
        ListNode* step2 = head;
        
        while (step2 && step2->next) {
            step1 = step1->next;
            step2 = step2->next->next;
            if (step1 == step2) {
                return true;
            }
        }
        return false;
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        seekEnd = head
        length = 0
        while seekEnd:
            seekEnd = seekEnd.next
            length += 1
        if length == n:
            return head.next
        currNode = head
        for _ in range(length - n - 1):
            currNode = currNode.next
        currNode.next = currNode.next.next
        return head
